### PR TITLE
RoPE: Rotary Positional Embeddings on attention layers (all 4 datasets)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current model uses no explicit positional encoding on attention layers — point coordinates are injected as input features but the transformer attention mechanism has no notion of relative geometry. **Rotary Positional Embeddings (RoPE)** encode relative position directly into the query-key dot product, which has shown strong gains in transformers operating on irregular spatial data (e.g. 3D point clouds).

For CFD surrogates, spatial locality is physically meaningful: nearby points interact more strongly through pressure gradients and boundary layer effects. RoPE gives the attention mechanism a learnable geometric inductive bias that vanilla attention lacks.

**Hypothesis**: Adding RoPE to the attention QK computation (encoding the 3D coordinates x/y/z as rotary angles) will improve surface pressure and field predictions across all datasets by allowing the model to reason about relative spatial distances between mesh nodes.

## Implementation

In `target/icml2026/train.py` (or the model definition), modify the multi-head attention layers:

1. For each attention layer, after computing Q and K projections, apply rotary embeddings using the node's (x, y, z) coordinates as the rotation angles:
   ```python
   # Standard RoPE for 3D: partition Q/K head dims into 3 pairs, rotate each by (theta_x, theta_y, theta_z)
   # Or use a 1D RoPE on a learned projection of xyz coordinates
   ```

2. Simplest implementation: use a **learned linear projection** of (x, y, z) → d_model/2 scalar angles, then apply 2D RoPE rotations. Alternatively use `xformers` or `rotary-embedding-torch` if available.

3. Apply RoPE **per layer** (all attention layers share the same RoPE scheme, different learned projections per layer is also fine).

4. Keep all other hyperparameters identical to the current best config per dataset.

**Suggested flag**: `--rope-dim 32` (number of RoPE embedding dimensions, must be even; try 32 as default). If a flag doesn't exist, implement it with this default.

### TandemFoil
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb_group rope-embeddings
```

### AirfRANS
```bash
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --wandb_group rope-embeddings
```

### DrivAerML
```bash
SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb_group rope-embeddings
```

### TandemFoil Paper
```bash
python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb_group rope-embeddings
```

## Current Baselines (targets to beat)

| Dataset | Metric | Current Best | PR |
|---------|--------|-------------|-----|
| TandemFoil | `val_primary/surface_pressure_mae` | **26.06** | #2887 |
| AirfRANS | `val_primary/surface_mse` | **0.000627** | #2902 |
| DrivAerML | `val_primary/surface_rel_l2_pct` | **3.997%** | #2898 |
| TandemFoil Paper | `val_primary/field_mse` | *no baseline yet* | — |

## Expected Outcome

RoPE has consistently improved spatial transformer models in point cloud and mesh settings. Expected improvement: 1-3% on surface pressure metrics across TF and AF. DrivAerML with its large 3D geometry may benefit the most from explicit rotational geometry encoding.

## Prior Results to Note

- Spike's prior PR #2901 (Huber loss) was sent back; Huber is now covered by Wave 3 (#2966 gohan), so this is a fresh direction
- No RoPE experiments have been attempted on this codebase

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT spike:

## Results

### RoPE (dim=32) Centroid Positional Embeddings — Cross-Dataset

**Hypothesis:** Applying Rotary Positional Embeddings to the Transolver's attention layers will improve spatial awareness and reduce prediction error across all datasets.

**Implementation:** RoPE is applied to Q/K vectors in each TransolverAttention layer using **slice centroid positions** — weighted spatial averages of node coordinates per slice token. This adapts RoPE (normally for discrete token positions) to the Transolver's slice-based attention where tokens are abstract aggregates rather than individual spatial points. Coordinates are threaded from the model input through the full backbone chain.

### Metrics Summary

| Dataset | Metric | Baseline | RoPE-32 Best | Epoch | Delta |
|---|---|---|---|---|---|
| AirfRANS | surface_mse | **0.000627** | 0.00573 | 87/90 | **+813% (9.1x worse)** |
| TandemFoil | surface_pressure_mae | **26.06** | 67.14 | 10/10 | **+158% (2.6x worse)** |
| DrivAerML | surface_rel_l2_pct | **3.997%** | 8.82% | 68/71 | **+121% (2.2x worse)** |
| TandemFoil Paper | field_mse | N/A | 0.0248 | 33/33 | No baseline |

**Note:** TandemFoil is only at epoch 10 (very slow due to 30-min torch.compile per launch). The baseline reached 26.06 at epoch 300. At similar early epoch counts, the RoPE trajectory (165→133→115→103→93→88→83→76→69→67) is still improving and may eventually plateau closer to baseline. However, the clear regressions on AirfRANS (90 epochs, 6+ cosine troughs) and DrivAerML (71 epochs, 2+ cosine troughs) are strong evidence that RoPE via centroids hurts.

### Per-Dataset Details

**AirfRANS** (W&B: `qn46wubf`, GPU 0, ~39 GiB peak)
- Best surface_mse: 0.00573 at ep87 (baseline 0.000627, external target 0.0043)
- Best volume_mse: 0.0365 at ep86
- Metric oscillates 0.006–0.034 across cosine cycles (T_max=10)
- 90 epochs completed, clear plateau

**TandemFoil** (W&B: `24pypzzr`, GPU 1, ~33 GiB peak)
- Best surface_pressure_mae: 67.14 at ep10 (baseline 26.06)
- Per-split at ep10: cruise=50.72, racecar=81.80, re_rand=65.88, single_in_dist=70.16
- Only 10 epochs — still improving, but trajectory suggests won't match baseline

**DrivAerML** (W&B: `dmly303i`, GPU 2, ~6 GiB active)
- Best surface_rel_l2_pct: 8.82% at ep68 (baseline 3.997%)
- 71 epochs completed, oscillating 8.8–14.4% across T_max=30 cycles
- Metric still slowly improving at cosine troughs

**TandemFoil Paper** (W&B: `g3zd26b5`, GPU 3, ~43 GiB peak)
- Best field_mse: 0.0248 at ep33 (no established baseline)
- Best surface_mse: 0.0067 at ep33
- Still improving steadily

### Commands Used

```bash
# AirfRANS
CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --rope-dim 32 --agent spike --wandb-name "spike/af-rope32" --wandb-group rope-embeddings

# TandemFoil
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --rope-dim 32 --agent spike --wandb-name "spike/tf-rope32" --wandb-group rope-embeddings

# DrivAerML
CUDA_VISIBLE_DEVICES=2 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --rope-dim 32 --agent spike --wandb-name "spike/dam-rope32" --wandb-group rope-embeddings

# TandemFoil Paper
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --rope-dim 32 --agent spike --wandb-name "spike/tfp-rope32" --wandb-group rope-embeddings
```

### What Happened

**RoPE via slice centroids is a clear negative result across all datasets.**

The fundamental issue is the mismatch between RoPE's design assumptions and the Transolver architecture:

1. **Centroid positions are too coarse.** Standard RoPE works because each token has a precise spatial/sequential position. In the Transolver, attention operates on "slice tokens" — abstract aggregates of many mesh nodes with diffuse spatial extent. The weighted centroid is a poor summary of "where" a slice is, losing the spatial resolution RoPE needs to encode meaningful relative positions.

2. **Rotation fights learned groupings.** The Transolver learns to cluster nodes into slices based on physical flow patterns (boundary layers, wakes, pressure gradients), not spatial proximity. RoPE biases attention toward spatially close slices via its distance-decay property, which conflicts with the physically-motivated groupings the model learns.

3. **Spatial information is already available.** The model receives raw coordinates via input embeddings and Fourier features. RoPE adds a second, potentially conflicting spatial signal in the attention mechanism, creating interference rather than reinforcement.

4. **The rotation itself adds noise.** With only 32 RoPE dimensions out of the full head dimension, the rotated portion of Q/K dot products introduces a positional bias that the model hasn't learned to use effectively, while the un-rotated portion continues to carry the learned attention patterns. This hybrid dilutes attention quality.

### Suggested Follow-ups

1. **Node-level RoPE before slice aggregation:** Instead of applying RoPE at the slice level, apply it to the raw node-level Q/K before the soft-clustering step. This would give each mesh node a precise position encoding. However, this requires significant architecture changes since the current design computes QKV after slice aggregation.

2. **Learned positional bias:** Instead of the rigid RoPE rotation, add a learned relative position bias to the slice attention scores (similar to ALiBi or relative attention). This would let the model discover whether spatial proximity matters for each head, rather than forcing it.

3. **Per-node sinusoidal injection:** Add Fourier-based positional encoding directly to node features before slice aggregation, separate from the existing Fourier features. This would be additive rather than multiplicative (unlike RoPE), avoiding interference with learned attention patterns.

4. **RoPE on a standard (non-slice) attention variant:** Test RoPE on a traditional per-node attention transformer to isolate whether the issue is RoPE itself vs the centroid approximation. If RoPE helps per-node attention, it confirms the centroid is the bottleneck.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2975 because ADVISOR: Closing. RoPE via slice centroids fails across all datasets.

AirfRANS: 0.00573 (9.1x worse than baseline 0.000627)
TandemFoil: 67.14 at ep10 — trajectory 165→67 improving but AirfRANS/DM results make the direction untenable
DrivAerML: 8.82% (2.2x worse than 3.997% baseline)

The core problem: Slice tokens in Transolver are soft semantic aggregates of mesh nodes, not individual spatial points with well-defined positions. Using the weighted centroid of node coordinates as the 'position' for RoPE is geometrically meaningful for individual nodes but creates a noisy, non-monotonic position signal for slice tokens — slices can overlap spatially, and their centroids may not encode the information RoPE needs to work well.

RoPE is designed for contexts where position has a clear sequential or spatial meaning (tokens in a sentence, pixels in an image). The Transolver's slice-based attention breaks that assumption. The idea is fundamentally mismatched to this architecture.

Note: PR #2983 (zenitsu) is also running RoPE — that should be terminated or redirected when it checks in.

No follow-up warranted on RoPE for this architecture..
